frontend/src/api/index.ts

```diff
@@ -10,13 +11,36 @@ let api: Client | undefined;
 let waiters: Array<Waiter> = [];
 
 export default {
-  init: () => {
+  init: async (actions) => {
     api = new Client(METABACKEND_URL, METABACKEND_NETWORK);
-    api.initialize().then(()=>{
-        // Wake up sleepers waiting for API.
-        for (let i = 0; i < waiters.length; i++) {
-          waiters[i](api);
+    (async () => {
+      try {
+        await api.initialize();
+        actions.afterSuccessfulInitialization();
+      } catch (e) {
+        switch (e.constructor) {
+          case ProviderLocked:
+            console.log("Metamask locked.")
+            actions.metamaskLocked();
+            break;
+          case ProviderOnUnsupportedNetwork:
+            console.log("Metamask on unsupported network.")
+            actions.unsupportedNetwork();
+            break;
+          case ProviderInitializeError:
+            console.log("Metamask error (is it installed?): " + e);
+            actions.metamaskNotInstalled();
+            break;
+          default:
+            console.log("Unknown error: " + e);
+            actions.unknownError()
         }
+      }
+    })().then(() => {
```

Comment:
you already have a linear `async` function which caches exceptions -> `then` is always gonna be called. No need to use mulitple paradigms at once with async and promises in the same, linear code.